beacon-chain/state/state-native/getters_participation.go

```diff
@@ -41,21 +48,67 @@ func (b *BeaconState) PreviousEpochParticipation() ([]byte, error) {
 // UnrealizedCheckpointBalances returns the total balances: active, target attested in
 // current epoch and target attested in previous epoch. This function is used to
 // compute the "unrealized justification" that a synced Beacon Block will have.
-func (b *BeaconState) UnrealizedCheckpointBalances() (uint64, uint64, uint64, error) {
-	if b.version == version.Phase0 {
-		return 0, 0, 0, errNotSupported("UnrealizedCheckpointBalances", b.version)
-	}
-
+func (b *BeaconState) UnrealizedCheckpointBalances(ctx context.Context) (uint64, uint64, uint64, error) {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	cp := b.currentEpochParticipation
-	pp := b.previousEpochParticipation
-	if cp == nil || pp == nil {
-		return 0, 0, 0, ErrNilParticipation
-	}
+	var cp, pp []byte
 
 	currentEpoch := time.CurrentEpoch(b)
+	if b.version == version.Phase0 {
+		targetIdx := params.BeaconConfig().TimelyTargetFlagIndex
```

Comment:
is  it possible to abstract all of this into a common method that can be re-used both by the proto and native state ? Also this particular part of the code base engages in recursive read locking, which can lead to deadlocks in certain cases. `helpers.BlockRoot` will acquire the read lock again.